packages/node/src/indexer/BlockedQueue.ts

```diff
@@ -54,3 +54,78 @@ export class BlockedQueue<T> {
     return result;
   }
 }
+
+export class Queue<T> {
+  private items: T[] = [];
+
+  get size(): number {
+    return this.items.length;
+  }
+
+  put(item: T): void {
+    this.putMany([item]);
+  }
+
+  putMany(items: T[]): void {
+    this.items.push(...items);
+  }
+
+  take(): T | undefined {
+    return this.items.shift();
+  }
+
+  takeAll(): T[] {
+    const result = this.items;
+
+    this.items = [];
+    return result;
+  }
+}
+
+type Action<T> = {
+  task: () => Promise<T> | T;
+  resolve: (value: T) => void;
+  reject: (reason: any) => void;
+};
+
+export class AutoQueue<T> {
+  private pendingPromise = false;
+  private queue = new Queue<Action<T>>();
+
+  async put(item: () => T): Promise<T> {
+    return this.putMany([item])[0];
+  }
+
+  putMany(tasks: Array<() => T>): Promise<T>[] {
+    return tasks.map((task, index) => {
+      return new Promise((resolve, reject) => {
+        this.queue.put({ task, resolve, reject });
+        if (tasks.length - 1 === index) {
+          void this.take();
+        }
+      });
+    });
+  }
+
+  async take(): Promise<boolean> {
```

Comment:
Do i understand it correct, that when a batch of task is pushed into the queue, `take()` will be called same times initially, but except the first call, you expect the others will return false and exit early. The first success call will then recursively call itself until all tasks in the queue are executed.

Reply:
`take` gets called again to ensure that we are processing items in the queue because it stops being recursive if the queue is emptied. It only needs to be called a single time for a batch in order to start processing again.

Its based on this answer https://stackoverflow.com/a/63208885/7492768 which is well worth the read